src/Compilers/VisualBasic/Portable/Parser/BlockContexts/SingleLineElseContext.vb

```diff
@@ -92,34 +93,90 @@ EndBlock:
+
+        Friend Overrides Function ResyncAndProcessStatementTerminator(statement As StatementSyntax, lambdaContext As BlockContext) As BlockContext
+            Dim token = Parser.CurrentToken
+            Select Case token.Kind
+                Case SyntaxKind.StatementTerminatorToken, SyntaxKind.EndOfFileToken, SyntaxKind.ColonToken
+                    Return ProcessStatementTerminator(lambdaContext)
+
                 Case SyntaxKind.ElseKeyword
-                    ' Terminated by Else from containing block.
-                    Parser.ConsumedStatementTerminator(allowLeadingMultilineTrivia:=False)
-                    Return ProcessElseAsStatementTerminator()
+                    If TreatElseAsStatementTerminator Then
+                        ' Terminated by Else from containing block.
+                        Parser.ConsumedStatementTerminator(allowLeadingMultilineTrivia:=False)
+                        Return ProcessElseAsStatementTerminator()
+                    End If
+
+                    Return MyBase.ResyncAndProcessStatementTerminator(statement, lambdaContext)
 
                 Case Else
                     ' Terminated if we've already seen at least one statement.
                     If _statements.Count > 0 Then
-                        Return ProcessOtherAsStatementTerminator()
-                    End If
-                    ' Terminated if the next token can follow a statement.
-                    If Parser.CanFollowStatementButIsNotSelectFollowingExpression(token) Then
-                        Return ProcessOtherAsStatementTerminator()
+                        If TreatOtherAsStatementTerminator Then
+                            Return ProcessOtherAsStatementTerminator()
+                        End If
+
+                        Return MyBase.ResyncAndProcessStatementTerminator(statement, lambdaContext)
                     End If
+
                     Parser.ConsumedStatementTerminator(allowLeadingMultilineTrivia:=False)
                     Return Me
-
             End Select
         End Function
 
-        Friend Overrides Function ProcessElseAsStatementTerminator() As BlockContext
-            Dim context = EndBlock(Nothing)
-            Return context.ProcessElseAsStatementTerminator()
-        End Function
-
-        Friend Overrides Function ProcessOtherAsStatementTerminator() As BlockContext
-            Dim context = EndBlock(Nothing)
-            Return context.ProcessOtherAsStatementTerminator()
+        Private ReadOnly Property TreatElseAsStatementTerminator As Boolean
+            Get
+                ' We can treat 'Else' as a valid statement terminator only
+                ' if there is a line-If without Else up the block context chain, or
+                ' if we are inside a single-line statement lambda and the 'Else' terminates the lambda.
+                Debug.Assert(BlockKind = SyntaxKind.SingleLineElseClause)
+                Debug.Assert(PrevBlock.BlockKind = SyntaxKind.SingleLineIfStatement)
```

Comment:
The property can always be invoked, this asserts an invariant in the way block contexts are nested for line-if. They are always nested this way.